osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using Amazon;
+using Amazon.Runtime;
+using Amazon.S3;
+using Amazon.S3.Model;
+using osu.Framework.Extensions;
+using osu.Game.IO.Archives;
+using osu.Server.BeatmapSubmission.Configuration;
+using osu.Server.BeatmapSubmission.Models;
+using SharpCompress.Writers;
+using SharpCompress.Writers.Zip;
+
+namespace osu.Server.BeatmapSubmission.Services
+{
+    public class S3BeatmapStorage : IBeatmapStorage
+    {
+        private const string osz_directory = "osz";
+        private const string versioned_file_directory = "beatmap_files";
+
+        private readonly AmazonS3Client client;
+
+        public S3BeatmapStorage()
+        {
+            client = new AmazonS3Client(
+                new BasicAWSCredentials(AppSettings.S3AccessKey, AppSettings.S3SecretKey),
+                new AmazonS3Config
+                {
+                    CacheHttpClient = true,
+                    HttpClientCacheSize = 32,
+                    RegionEndpoint = RegionEndpoint.USWest1,
+                    UseHttp = true,
+                    ForcePathStyle = true,
+                    RetryMode = RequestRetryMode.Legacy,
+                    MaxErrorRetry = 5,
+                    Timeout = TimeSpan.FromMinutes(1),
+                });
+        }
+
+        public async Task StoreBeatmapSetAsync(uint beatmapSetId, byte[] beatmapPackage)
+        {
+            var stream = new MemoryStream(beatmapPackage);
+            using var archiveReader = new ZipArchiveReader(stream);
+
+            var files = new List<byte[]>();
+
+            foreach (string filename in archiveReader.Filenames)
+                files.Add(await archiveReader.GetStream(filename).ReadAllBytesToArrayAsync());
+
+            await Task.WhenAll(
+                uploadBeatmapPackage(beatmapSetId, beatmapPackage, stream),
+                uploadBeatmapFiles(files));
```

Comment:
There is some parallelisation here because in testing S3 was being pretty slow for me.

Note that due to the structure of everything every beatmap upload basically incurs a full S3 roundtrip, e.g. a download of the previous package and an upload of the new package, plus the upload of the individual files.

The files upload is a bit wasteful, since there is a possibility of the files already existing. However, eliminating that is annoying structurally; I'd have to pass in stuff to indicate which files to upload and which to not which gets a bit dicey for me. It can probably be done, though.

Reply:
It's always good to have parallel uploads for S3, because yeah, the round trip time can fluctuate (especially based around server region). I think this stems from the upload performing quite a few requests to get to the final state.